admin/code/LeftAndMain.php

```diff
@@ -356,37 +355,40 @@ protected function getSchemaResponse($form, $id = null) {
 	 * Returns a representation of the provided {@link Form} as structured data,
 	 * based on the request data.
 	 *
-	 * @param Form $form
-	 * @param String $id Optional, will default to the current request URL
+	 * @param string $schemaID ID for this schema. Required.
+	 * @param Form $form Required for 'state' or 'schema' response
+	 * @param ValidationResult $errors Required for 'error' response
 	 * @return array
 	 */
-	protected function getSchemaForForm(Form $form, $id = null) {
-		$request = $this->getRequest();
-		$id = $id ? $id : $request->getURL();
-		$return = null;
-
-		// Valid values for the "X-Formschema-Request" header are "schema" and "state".
-		// If either of these values are set they will be stored in the $schemaParst array
-		// and used to construct the response body.
-		if ($schemaHeader = $request->getHeader('X-Formschema-Request')) {
-			$schemaParts = array_filter(explode(',', $schemaHeader), function($value) {
-				$validHeaderValues = ['schema', 'state'];
-				return in_array(trim($value), $validHeaderValues);
-			});
-		} else {
-			$schemaParts = ['schema'];
-		}
+	protected function getSchemaForForm($schemaID, Form $form = null, ValidationResult $errors = null) {
+		$schemaHeader = $this->getRequest()->getHeader('X-Formschema-Request');
+		$schemaParts = preg_split('#\s*,\s*#', $schemaHeader) ?: [];
+		$wantSchema = in_array('schema', $schemaParts);
+		$wantState = in_array('state', $schemaParts);
+		$wantErrors = in_array('errors', $schemaParts);
+		$auto = in_array('auto', $schemaParts);
 
-		$return = ['id' => $id];
+		// Require ID
+		if (empty($schemaID)) {
+			throw new InvalidArgumentException("schemaID is required");
+		}
+		$return = ['id' => $schemaID];
 
-		if (in_array('schema', $schemaParts)) {
```

Comment:
I wonder if this logic should be refactored out of `LeftAndMain` and into `FormSchema`, as it's likely to be a part of any interaction between a rich JS form and a supporting controller.

Something like this:

```php
return $this->schema->getMultipartSchema($schemaID, $schemaParts, $form, $errors);
```

Also I think that 'result' would be a better name than 'errors' for component generated by the ValidationResult, as a ValidationResult might also say "valid".